substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+/// This keeps track of the active balance of the delegatee that is made up from the funds that are
+/// currently delegated to this delegatee. It also tracks the pending slashes yet to be applied
+/// among other things.
+#[derive(Default, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct DelegationLedger<T: Config> {
+	/// Where the reward should be paid out.
+	pub payee: T::AccountId,
+	/// Sum of all delegated funds to this delegatee.
+	#[codec(compact)]
+	pub total_delegated: BalanceOf<T>,
+	/// Amount that is bonded and held.
+	// FIXME(ank4n) (can we remove it)
+	#[codec(compact)]
+	pub hold: BalanceOf<T>,
+	/// Slashes that are not yet applied.
+	#[codec(compact)]
+	pub pending_slash: BalanceOf<T>,
+	/// Whether this delegatee is blocked from receiving new delegations.
+	pub blocked: bool,
+}
+
+impl<T: Config> DelegationLedger<T> {
+	/// balance that can be staked.
+	pub fn delegated_balance(&self) -> BalanceOf<T> {
+		// do not allow to stake more than unapplied slash
+		self.total_delegated.saturating_sub(self.pending_slash)
+	}
+
+	/// balance that is delegated but not bonded.
+	pub fn unbonded_balance(&self) -> BalanceOf<T> {
+		self.total_delegated.saturating_sub(self.hold)
+	}
+}
+
+impl<T: Config> DelegationInterface for Pallet<T> {
+	type Balance = BalanceOf<T>;
+	type AccountId = T::AccountId;
+
+	fn delegated_balance(who: &Self::AccountId) -> Self::Balance {
+		<Delegatees<T>>::get(who)
+			.map_or_else(|| 0u32.into(), |register| register.delegated_balance())
+	}
+
+	fn unbonded_balance(who: &Self::AccountId) -> Self::Balance {
+		<Delegatees<T>>::get(who)
+			.map_or_else(|| 0u32.into(), |register| register.unbonded_balance())
+	}
+
+	fn accept_delegations(
+		who: &Self::AccountId,
+		reward_destination: &Self::AccountId,
+	) -> DispatchResult {
+		// Existing delegatee cannot register again.
+		ensure!(!<Delegatees<T>>::contains_key(who), Error::<T>::NotAllowed);
+
+		// A delegator cannot become a delegatee.
+		ensure!(!<Delegators<T>>::contains_key(who), Error::<T>::NotAllowed);
+
+		// payee account cannot be same as delegatee
+		ensure!(reward_destination != who, Error::<T>::InvalidRewardDestination);
+
+		// make sure they are not already a direct staker or they are migrating.
+		ensure!(
+			T::CoreStaking::status(who).is_err() || <DelegateeMigration<T>>::contains_key(who),
+			Error::<T>::AlreadyStaker
+		);
+
+		// already checked delegatees exist
+		<Delegatees<T>>::insert(
+			who,
+			DelegationLedger {
+				payee: reward_destination.clone(),
+				total_delegated: Zero::zero(),
+				hold: Zero::zero(),
+				pending_slash: Zero::zero(),
+				blocked: false,
+			},
+		);
+
+		Ok(())
+	}
+
+	/// Transfers funds from current staked account to `proxy_delegator`. Current staked account
+	/// becomes a delegatee with `proxy_delegator` delegating stakes to it.
+	fn migrate_accept_delegations(
+		new_delegatee: &Self::AccountId,
+		proxy_delegator: &Self::AccountId,
```

Comment:
If any account can be provided as a `proxy_delegator`, someone could pluck out a random account from Subscan, use it as the proxy delegator, and block them from being a nominator / validator.

Think we discussed the idea of having this being an account that is created on the fly in a similar way to the way pools does it. We'd probably need a counter, e.g. `LastMigrationAccountId`, and have the caller pay the ED in this account, that could be released once the migration is over and there is only ED left in the account.

Does this approach sound viable?